File: dotnet/Data/TodoRepository.cs
using Microsoft.EntityFrameworkCore;
using Todo.Core.Abstract.Repositories;
using Todo.Core.Models;
using Todo.Data.Mappers;

namespace Todo.Data;

internal class TodoRepository(TodoDbContext context) : ITodoRepository
{
    public async Task<TodoEntity> Create(TodoEntity entity)
    {
        var data = entity.ToData();

        await context.Set<Models.Todo>().AddAsync(data);
        await context.SaveChangesAsync();

        return data.ToEntity();
    }

    public async Task<TodoEntity?> Get(int id)
    {
        var data = await context.Set<Models.Todo>()
            .AsNoTracking()
            .Where(t => !t.IsDone)
            .Where(t => t.Id == id)
            .FirstOrDefaultAsync();

        return data?.ToEntity();
    }

    public async Task<ICollection<TodoEntity>> GetAll()
    {
        var data = await context.Set<Models.Todo>()
            .AsNoTracking()
            .Where(t => !t.IsDone)
            .ToArrayAsync();

        return data.ToEntities();
    }

    public async Task<TodoEntity> Update(TodoEntity entity)
    {
        var data = entity.ToData();

        context.Set<Models.Todo>().Update(data);
        await context.SaveChangesAsync();

        return data.ToEntity();
    }

    public async Task<bool> Delete(int id) => await context.Set<Models.Todo>()
        .Where(t => t.Id == id)
        .ExecuteDeleteAsync() != 0;
}

File: nodejs/src/data/todo-repository.ts
import { Repository } from "typeorm";
import { ITodoRepository } from "../core/abstract/repositories/todo-repository";
import { Todo } from "../core/todo";

export class TodoRepository implements ITodoRepository {
  constructor(private readonly repository: Repository<Todo>) {}

  async create(todo: Todo): Promise<Todo> {
    const saved = await this.repository.save([todo]);

    if (saved.length != 1)
      throw new Error("Unexpected error during saving todo");

    return saved[0];
  }

  getAll(): Promise<Todo[]> {
    return this.repository.find({ where: { isDone: false } });
  }

  async delete(id: number): Promise<boolean> {
    const result = await this.repository.update({ id }, { isDone: true });

    return result.affected == 1;
  }
}
